86 postgres replica (#168)
open-contracting/deploy: docs/deploy/create_server.rst

#. Go to `Hetzner <https://www.hetzner.com/?country=us>`__
#. Click "Dedicated", and navigate to choose a suitable server for your application.
#. Click the "Order" button for the server that you've chosen

   #. Select a location; we've never had an issue with simply choosing the cheapest
   #. Select an operating system - "Ubuntu 18.04 LTS minimal"
   #. Select any additional storage required

#. Click "Order Now"
#. (optionally) Select "Public key" in the "Server Login Details" section, and paste your SSH key in; this will be added to /root/.ssh/authorized_keys
#. Click "Save"
#. Review the contents of the cart, then click "Checkout"
#. Log in using OCP's credentials. This will happen automatically if you're already logged into Hetzner services.
#. Check the "I have read your Terms and Conditions as well as your Privacy Policy and I agree to them." box
#. Click "Order in Obligation"
#. Wait until you receive an email notifying you that your server is ready, then proceed to deploying the service.

> Some Hetzner servers only let you start on their recovery OS.
> 
> If you were not able to select Ubuntu above, you will need to follow these additional steps:

Reviewer: ```suggestion
If you couldn't select Ubuntu above, follow these additional steps:
```

#. SSH into recovery image

> #. Test the server hardware:

Reviewer: ```suggestion
#. Test the server hardware. The SMART values to consider vary depending on the RAID controller and disk manufacturer. Ask a colleague if you need help.
```

Reviewer: Thanks!

   .. code-block:: bash

      smartctl -t long /dev/<device>
      smartctl -a /dev/<device>

> #. Run the pre-installed `Hetzner OS installer <https://github.com/hetzneronline/installimage>`__ and accept the defaults, unless stated otherwise below:

Reviewer: ```suggestion
#. Run the pre-installed `Hetzner OS installer <https://github.com/hetzneronline/installimage>`__ (`see documentation <https://wiki.hetzner.de/index.php/Installimage/en>`__) and accept the defaults, unless stated otherwise below:
```

   .. code-block:: bash

      installimage

   #. Select "Ubuntu 18.04 - minimal"

   #. The installer opens a configuration file with a number of install options.

>       #. Set ``DRIVE1``, ``DRIVE2``, etc. to the physical disks you want

Reviewer: ```suggestion
      #. Set ``DRIVE1``, ``DRIVE2``, etc. to the drives you want to use. (You can identify drives with the ``smartctl`` command.) If you ordered two large drives for a server that already includes two small drives, you might only set the large drives. For example:

         .. code-block:: code

            DRIVE1=/dev/sdb
            DRIVE2=/dev/sdd
```

      #. Set ``SWRAIDLEVEL 1``

>       #. Set ``HOSTNAME <hostname>``

Reviewer: ```suggestion
      #. Set the hostname. For example:

      .. code-block:: none

         HOSTNAME example.open-contracting.org

```

      #. Create partitions. Set the ``swap`` partition size according to the comments in `swap.sls <https://github.com/open-contracting/deploy/blob/master/salt/core/swap.sls>`__. For example:

>       .. code-block:: none
> 
>          PART swap swap 16G
>          PART /boot ext2 1G
>          PART / ext4 all

Reviewer: Fix my indentation:

```suggestion
         .. code-block:: none

            PART swap swap 16G
            PART /boot ext2 1G
            PART / ext4 all
```

   #. Press ``F2`` to save

>    #. Overwrite drives

Reviewer: ```suggestion
   #. Confirm that you want to overwrite the drives, when prompted
```

#. Click "Dedicated", and navigate to choose a suitable server for your application. So far, we've used EX-Line servers.
#. Click the "Order" button for the server that you've chosen

   #. Select a location; we've never had an issue with simply choosing the cheapest
   #. Select an operating system - "Ubuntu 18.04 LTS minimal"
   #. Select any additional storage required

#. Click "Order Now"
#. (optionally) Select "Public key" in the "Server Login Details" section, and paste your SSH key in; this will be added to /root/.ssh/authorized_keys
#. Click "Save"
#. Review the contents of the cart, then click "Checkout"
#. Log in using OCP's credentials. This will happen automatically if you're already logged into Hetzner services.
#. Check the "I have read your Terms and Conditions as well as your Privacy Policy and I agree to them." box
#. Click "Order in Obligation"
#. Wait until you receive an email notifying you that your server is ready, then proceed to deploying the service.

Some Hetzner servers only let you start on their recovery OS.

If you were not able to select Ubuntu above, you will need to follow these additional steps:

> #. SSH into recovery image

Author: It's the same as SSH'ing into the server. Perhaps this is clearer? 

> SSH into the server, which has booted into the recovery image.

> #. SSH into recovery image

Reviewer: ```suggestion
#. Activate and load the `Rescue System <https://wiki.hetzner.de/index.php/Hetzner_Rescue-System/en>`__, if not already loaded.
#. Connect to the server as the ``root`` user using the password provided when activating the Rescue System.
```

#. Test the server hardware

  .. code-block:: bash

    smartctl -t long /dev/<device>

>     smartctl -a /dev/<device>

Author: I was trying to make this section a prompt to test the hardware rather than an exact checklist.

Due to differences in hardware between servers, even with the same supplier, it's tricky to write a full testing checklist.

For example, with disks, the SMART values vary between makes, we are checking in Prometheus for specific Samsung values that won't exist on other disks. 

There are other bits of hardware to test as well, for instance hardware raid (if we start a server with it), but again this is hugely different depending on the RAID controller. 
If you were being very thorough you could run a full memory test - I wouldn't recommend this unless we suspected issues with the server memory as it takes hours/days to run.

To some extent we should trust Hetzner to not give us bad hardware equally now is the best time for issues to be identified. 

Does this make sense? 
How do you think it is best for the docs to reflect this?

#. Run the pre-installed `Hetzner OS installer <https://github.com/hetzneronline/installimage>`

   .. code-block:: bash

      installimage

   #. Follow the default installer prompts, unless specified.

   #. Select Ubuntu 18.04 - minimal

   #. The installer takes you to a configuration file with a number of install options.

      .. code-block:: none

>          Set DRIVE1 and DRIVE2 etc to the physical disks you want

Reviewer: These configs have an `=`, whereas others just put a space?

>          Set DRIVE1 and DRIVE2 etc to the physical disks you want

Author: I was wrong, they shouldn't have a `=`. 
I'm happy with your suggestion and have committed it.

         ...
         SWRAIDLEVEL 1
         ...

>          HOSTNAME <server hostname>

Author: I put in `replica1.kingfisher.open-contracting.org` 
Perhaps update the documentation to `example.open-contracting.org`

Ideally, we should link to page/section in the docs detailing out a hostname standard - I'm not sure if there is already a github issue about this?

>          HOSTNAME <server hostname>

Reviewer: I made a code-level suggestion. There's an issue for Salt target naming #115, but please open an issue for hostname naming.

>          HOSTNAME <server hostname>

Author: Issue raised #172

         ...
         PART swap swap 16G
         PART /boot ext2 1G
         PART / ext4 all

      For swap partition sizings refer to `salt coniguration<https://github.com/open-contracting/deploy/blob/master/salt/core/swap.sls>`.

>    #. F2 # Save

Author: Yes that's exactly right. I'll tweak it a bit to be clearer. 
> Press F2 to save your changes

>    #. Overwrite drives

Author: Overwrite drives is a prompt given by the installer.
It just double checking things are correct for software raid to overwrite the hard drives.

>    #. Overwrite drives

Author: Committed suggestion fixing this.

#. Click the "Order" button for the server that you've chosen

   #. Select a location; we've never had an issue with simply choosing the cheapest
   #. Select an operating system - "Ubuntu 18.04 LTS minimal"
   #. Select any additional storage required

#. Click "Order Now"
#. (optionally) Select "Public key" in the "Server Login Details" section, and paste your SSH key in; this will be added to /root/.ssh/authorized_keys
#. Click "Save"
#. Review the contents of the cart, then click "Checkout"
#. Log in using OCP's credentials. This will happen automatically if you're already logged into Hetzner services.
#. Check the "I have read your Terms and Conditions as well as your Privacy Policy and I agree to them." box
#. Click "Order in Obligation"
#. Wait until you receive an email notifying you that your server is ready, then proceed to deploying the service.

If you couldn't select Ubuntu above, follow these additional steps:

#. Activate and load the `Rescue System <https://wiki.hetzner.de/index.php/Hetzner_Rescue-System/en>`__, if not already loaded.
#. Connect to the server as the ``root`` user using the password provided when activating the Rescue System.

> #. Test the server hardware:
> 
>    .. code-block:: bash
> 
>       smartctl -t long /dev/<device>
>       smartctl -a /dev/<device>
> 

Author: ```suggestion
#. Test the server hardware:
   #. Test the drives. The SMART values to look for vary depending on the drive manufacturer. Ask a colleague if you need help. 
   
   .. code-block:: bash

      smartctl -t long /dev/<device>
      smartctl -a /dev/<device>

   #. Test the hardware RAID controller, if there is one. The  software to do so varies based on the RAID controller. Ask a colleague if you need help.

```
